DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,769 @@
+
+## Contents
+* [Background theory](#background-theory)
+* [Rationale](#rationale)
+* [Prior work](#prior-work)
+* [Other languages](#other-languages)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reference](#reference)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Background theory
+Pure functions in programming languages can often be thought of as mathematical functions, mapping elements from an input set (domain) to elements of an output set (co-domain).
+A big problem with this comparison is that mathematical functions are defined to give an answer, while a computer has to perform computation [[1]](#reference).
+It is possible that a procedure just loops forever, and finding this out beforehand in general is impossible because of the [Halting Problem](https://en.wikipedia.org/wiki/Halting_problem).
+One solution is to not allow programs that may not halt.
+[Coq](https://en.wikipedia.org/wiki/Coq) does this, and consequently it's not turing complete.
+The more common solution is to have one implicit extra member to every type: the bottom value, denoted by `⊥` (called 'falsum').
+This is not a traditional value that is stored in memory with a certain bit-pattern, it simply represents the possibility that the code never reaches the point where an actual value of that type would be returned or assigned.
+
+For example:
+```D
+bool isPrime(int x);
+bool foo(int x) {
+    return isPrime(x);
+}
+```
+While a `bool` usually only can be `true` or `false` and requires 1 bit of storage, foo can actually return three possible things: {`true`, `false`, `⊥`}
+The symbol `⊥` represents `isPrime` getting stuck in an endless loop (in a bad implementation), crashing (e.g. because it's out of memory), or throwing an exception (maybe x was negative).
+In such cases foo does not return a `true` or `false` value like a mathematical function always would.
+Note that having a bottom *type* in a language does not *add* the bottom *value* or any run time cost associated with it, the bottom value is a concept that exists in any turing complete language even without a bottom type.
+
+The bottom type should not be confused with a unit type, such as `void`:
+```D
+void assertInBounds(int[] arr, int x);
+void foo(int[] arr, int x) {
+    return assertInBounds(arr, x);
+}
+```
+A unit type has only one value, so 0 bits of storage are needed.
+Calling `assertInBounds` can produce two things: {`()`, `⊥`}
+Either it succeeds or it causes a crash.
+Since the unit value carries no information, it is discarded and doesn't produce any code or use any memory.
+The name of the unit value that is used here `()` is chosen arbitrarily, it could also be named `null`, `std::monostate` (C++), `undefined` (JavaScript), `None` (Scala) or `void.init`.
```

Comment:
In Scala it's `Unit`

Reply:
Thanks. I removed the line since I don't think the other names are useful here.